### PR TITLE
[FLINK-26675][runtime] Parallelized heavy serialization in StreamingJ…

### PR DESCRIPTION
**Pull request by KarmaGYZ**
…obGraphGenerator

## What is the purpose of the change

Parallelized heavy serialization in StreamingJobGraphGenerator

## Brief change log

- Introduce a `serializationExecutor` to `StreamingJobGraphGenerator`
- Move the heavy serialization to the `serializationExecutor`

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? (not applicable

### CODE REVIEW

**Comment by KarmaGYZ** [comment]
@flinkbot run azure

**Comment by KarmaGYZ** [comment]
@flinkbot run azure

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 1]
```diff
@@ -144,6 +151,18 @@ public static JobGraph createJobGraph(StreamGraph streamGraph, @Nullable JobID j
     private final StreamGraphHasher defaultStreamGraphHasher;
     private final List<StreamGraphHasher> legacyStreamGraphHashers;
 
+    private final ExecutorService serializationExecutor =
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:154]: Where do you shut this down?

Comment (KarmaGYZ) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:154]: Thanks, I also realize it and will fix it in the next version.

**Comment by KarmaGYZ** [comment]
Thanks for the review and valuable comments @zentol . I'm still working on the next version of this PR and will take your comments into account.

**Comment by KarmaGYZ** [comment]
@flinkbot run azure

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java [hunk 2]
```diff
@@ -132,6 +136,12 @@ public class StreamConfig implements Serializable {
 
     private final Configuration config;
 
+    private final transient Map<String, Object> objectSerializationConfig = new HashMap<>();
```
Comment (xintongsong) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:139]:
```suggestion
    private final transient Map<String, Object> toBeSerializedConfigObjects = new HashMap<>();
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java [hunk 3]
```diff
@@ -132,6 +136,12 @@ public class StreamConfig implements Serializable {
...
+    private final transient Map<Integer, CompletableFuture<StreamConfig>> chainedTaskFutures =
+            new HashMap<>();
+    private final transient CompletableFuture<StreamConfig> serializationFuture =
+            new CompletableFuture<>();
```
Comment (xintongsong) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:143]: I'd suggest to add a brief explanation here:
- We first collect all the need-to-be-serialized objects, then serialize them at once
- The purpose is to make parallelization of `StreamConfig` serialization easier.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java [hunk 4]
```diff
@@ -140,6 +150,67 @@ public Configuration getConfiguration() {
...
+                        },
+                        ioExecutor);
+        return serializationFuture;
+    }
+
+    /**
+     * Serialize all object configs synchronously. Only used for operators which need to reconstruct
+     * the StreamConfig internally or test.
+     */
+    public void serializedAllConfigs() {
```
Comment (xintongsong) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:190]:
```suggestion
    public void serializeAllConfigs() {
```

**Review by xintongsong — APPROVED** [review]
Thanks @KarmaGYZ. LGTM. I have some minor suggestions regarding the readability.

**Pull request closed without merging**